controller/src/main/java/io/pravega/controller/server/bucket/RetentionBucketService.java

```diff
@@ -0,0 +1,84 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.server.bucket;
+
+import io.pravega.client.stream.impl.StreamImpl;
+import io.pravega.common.concurrent.Futures;
+import io.pravega.common.hash.RandomFactory;
+import io.pravega.common.tracing.RequestTracker;
+import io.pravega.controller.store.stream.BucketStore;
+import io.pravega.controller.store.stream.OperationContext;
+import io.pravega.controller.store.stream.StreamMetadataStore;
+import io.pravega.controller.task.Stream.StreamMetadataTasks;
+import io.pravega.controller.util.Config;
+import io.pravega.controller.util.RetryHelper;
+import io.pravega.common.tracing.TagLogger;
+
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.function.Supplier;
+import org.slf4j.LoggerFactory;
+
+public class RetentionBucketService extends AbstractBucketService {
+    private static final TagLogger log = new TagLogger(LoggerFactory.getLogger(RetentionBucketService.class));
+
+    private final StreamMetadataTasks streamMetadataTasks;
+    private final StreamMetadataStore streamMetadataStore;
+    private final RequestTracker requestTracker;
+    private final Supplier<Long> requestIdGenerator = RandomFactory.create()::nextLong;
+
+    public RetentionBucketService(int bucketId, StreamMetadataStore streamMetadataStore, BucketStore bucketStore,
+                           StreamMetadataTasks streamMetadataTasks, ScheduledExecutorService executor,
+                           RequestTracker requestTracker) {
+        super(BucketStore.ServiceType.RetentionService, bucketId, bucketStore, executor);
+        this.streamMetadataTasks = streamMetadataTasks;
+        this.streamMetadataStore = streamMetadataStore;
+        this.requestTracker = requestTracker;
+    }
+    
+    CompletableFuture<Void> startWork(StreamImpl stream) {
```

Comment:
This really isn't going to scale well to keep a separate task with notifications in memory for every stream in the system. We should find a way for this to work incrementally. Perhaps we could have an outer loop that process all the streams this worker owns and just read them in as needed.

Reply:
Yes, we had this same concern when we designed retention infrastructure. There is a separate scheduled job per stream. 
For large number of streams under a bucket, this means large memory footprint. 

One idea was to scan all the streams under a bucket and create an in-memory list. Then one outer loop scans the list in a tight loop and identifies which streams are ready for their next run and start the run. And to keep the footprint in check, start only a limited number of jobs (say `n` concurrent submissions to IO heavy jobs on a threadpool). So only `n` entries in the executor queue hence a limited memory footprint. 
The problem is if for some reason some jobs took long time to complete, they would cause delays in execution of jobs on unrelated streams. Esp if the delay occurs in more than `n` jobs. 
Also, maintaining "fixed" or near fixed periodicity of execution becomes harder with such a model. 
Especially if the periods between each run need to be small (like with watermarks). 
Basically we could maintain a tree-map of streams with last executed time and in next cycle submit next round of jobs while honouring the time period. 
But we could definitely improve upon the scalability by employing such a processing.

Reply:
It's true that if we processed the items serially, it would be a bottlekneck. However each one is asynchronous, and can be run in a thread pool. So we could easily have say 100 outstanding, in which case any one being slow wouldn't be a problem. And if all 100 were slow, the server is probably overloaded or down, in which case stalling might be the right thing to do. 

I think this sort of change will be easier once some of the other changes are made. So do those first.